### PR TITLE
Task 4

### PR DESCRIPTION
**Pull request by jegius**
- [x] 1 - Task 4.1 is implemented
- [x] 3 - TASK 4.2 is implemented lambda links are provided and returns data
- [x] 4 - TASK 4.3 is implemented lambda links are provided and products is stored in DB (call TASK 4.2 to see the product)
- [x] 5 - Your own Frontend application is integrated with product service (/products API) and products from product-service are 

represented on Frontend. Link to a working Front-End application is provided for cross-check reviewer. 

[link to frontend with integration](https://d20wes0bg8lspj.cloudfront.net)
[frontend pull request](https://github.com/jegius/nodejs-aws-fe/pull/3)
Additional (optional) tasks (but nice to have):
https://ungftsxv80.execute-api.eu-west-1.amazonaws.com/dev/products/available

- [x] +1 - POST/products lambda functions returns error 400 status code if product data is invalid
- [x] +1 - All lambdas return error 500 status code on any error (DB connection, any unhandled error in code)
- [x] +1 - All lambdas do console.log for each incoming requests and their arguments
- [ ] +1 - Transaction based creation of product (in case stock creation is failed then related to this stock product is not created

### CODE REVIEW

File: product-service/serverless.ts [hunk 1]
```diff
@@ -38,6 +38,18 @@ const serverlessConfiguration: Serverless = {
                 }
             ]
         },
+        addProduct: {
+            handler: 'handler.addProduct',
+            events: [
+                {
+                    http: {
+                        method: 'put',
```
Comment (MariiaTarasiuk) [inline product-service/serverless.ts]: the method must be post

File: product-service/handlers/products/addProduct/addProduct.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+import {ProductNotFound} from "../../../errors/errors";
+
+export const addProductHandler: APIGatewayProxyHandler = async ({body}) => {
+    try {
+        const requestData: Product = JSON.parse(body) as Product;
+        const [result] = await productDao.add({...emptyProduct, ...requestData}) as Product[]
+        return buildResponse(result);
+    } catch (error) {
+        return error instanceof ProductNotFound
+            ? buildResponse(error.message, ErrorCode.BAD_REQUEST)
```
Comment (MariiaTarasiuk) [inline product-service/handlers/products/addProduct/addProduct.ts:20]: not validate all field 
image

File: product-service/serverless.ts [hunk 3]
```diff
@@ -38,6 +38,18 @@ const serverlessConfiguration: Serverless = {
...
+                        path: 'products/available',
```
Comment (MariiaTarasiuk) [inline product-service/serverless.ts:52]: the path should be just "/products"

File: product-service/handlers/products/getProducts/getProducts.ts [hunk 4]
```diff
@@ -1,6 +1,15 @@
...
+import {productDao} from "../../../dao/product/productDao";
+import {ProductNotFound} from "../../../errors/errors";
 
-export const getAvailableProductsHandler: APIGatewayProxyHandler = async () => buildResponse(await getProducts())
+export const getAvailableProductsHandler: APIGatewayProxyHandler = async () => {
+    try {
+        return buildResponse(await productDao.getAll())
+    } catch (error) {
+        return error instanceof ProductNotFound
+            ? buildResponse(error.message, ErrorCode.BAD_REQUEST)
```
Comment (MishaHnatyshyn) [inline product-service/handlers/products/getProducts/getProducts.ts]: Do we need this in getAll() handler?

**Comment by MishaHnatyshyn** [comment]
Please, provide a link to Lambda endpoints.

**Comment by jegius** [comment]
@MishaHnatyshyn  done

**Comment by MariiaTarasiuk** [comment]
image
empty post return 200

**Pull request merged**